Automation point mouse wheel and double click (old)

tecknixia opened the pull request:
----
Edit: Replaced by #5291 and #5292  
----

Related to issue #5225 

Allows user to place cursor over an automation point in the Automation Editor, and use scroll wheel to adjust the y level for that point.

https://cdn.discordapp.com/attachments/332258319228207114/629248936099708928/AutomationEditor2-2019-10-03_04.21.20.mp4  (example, not current)

I didn't know where to put the variable, so I put it where I thought it might fit.  I'm still getting familiar with the codebase.  Also, I was wondering if "point" is a good term for this or not.  So please let me know if I should change something.

tecknixia committed:
Update AutomationEditor.h

tecknixia committed:
Update AutomationEditor.cpp

tecknixia committed:
Update AutomationEditor.h

tecknixia commented:
I just noticed that I didn't account for different number scales.  So, even if the increments are whole numbers, it still adjusts in increments of three decimal places.

Spekular commented:
I used "numPoints" in the automation pattern flipping PR, so "points"
sounds consistent to me.

On Tue, Oct 8, 2019, 07:34 tecknixia <notifications@github.com> wrote:

> I just noticed that I didn't account for different number scales. So, even
> if the increments are whole numbers, it still adjusts in increments of
> three decimal places.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/LMMS/lmms/pull/5232?email_source=notifications&email_token=ACEBLGWUTGWPG65DSBAL3Q3QNQLYFA5CNFSM4I6M4GD2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEAS2AOQ#issuecomment-539336762>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACEBLGWJPXXULHFQEWJNZ63QNQLYFANCNFSM4I6M4GDQ>
> .
>

tecknixia commented:
I believe I have created a fix for the decimal point step changes using:

`m_pattern->firstObject()->step<float>()`

Is there a way to update the pull request?  Do I close and create a new one?

tecknixia committed:
Update AutomationEditor.cpp

tecknixia commented:
I figured it out.

tecknixia committed:
Merge remote-tracking branch 'upstream/master'
Updating forked repo tecknixia/lmms

tecknixia commented:
Seems to work fine.

Only quirky behavior that I've noticed so far is that it does not lock the Y axis while over a point.  So if the scale for zoom for Y is larger than the window, the mouse will adjust the Y value for the point, and scroll the information in the window up or down at the same time.  Slightly confusing at first, but still practical.

CYBERDEViLNL commented on src/gui/editors/AutomationEditor.cpp:
```diff
@@ -1709,6 +1718,61 @@ void AutomationEditor::wheelEvent(QWheelEvent * we )
+					
+					if (pos_ticks < 0) {
+						pos_ticks = 0;
+					}
```
This is fine but could be replaced with: 
```cpp
pos_ticks = (pos_ticks < 0) ? 0 : pos_ticks;
```

tecknixia commented on src/gui/editors/AutomationEditor.cpp:
```diff
@@ -1105,7 +1110,11 @@ inline void AutomationEditor::drawCross( QPainter & p )
+			QToolTip::showText( tt_pos, QString::number( scaledLevel ), this );
+		} else if (m_pointYLevel != 0) {
+			QToolTip::showText( tt_pos, QString::number( m_pointYLevel ), this );
+		}
```
I wasn't sure about that... simply because the wiki also states:
"The majority of LMMS conventions follows the Qt Coding Style"
and according to Qt Coding:
"Use attached braces: The opening brace goes on the same line as the start of the statement. If the closing brace is followed by another keyword, it goes into the same line as well"

`// Wrong
 if (codec)
 {
 }
 else
 {
 }

 // Correct
 if (codec) {
 } else {
 }`

I had interpreted the LMMS coding conventions to suggest that you should use brackets as opposed to non-explicit blocking:

`// BAD - Example of not explicit blocking
if (m_sample > 0)
	--m_sample;`

Seems like a contradiction of conventions, perhaps some clarification in the wiki would be ideal.

tecknixia replied:
I suppose the combined use of the words "explicit blocking" and "Block brackets SHOULD be on their own line" (I misinterpreted) is clear enough to suggest that brackets should not start on the beginning line of the statement.  With the exception of the whole statement fitting on one line, of course.

tecknixia commented on src/gui/editors/AutomationEditor.cpp:
```diff
@@ -1709,6 +1718,61 @@ void AutomationEditor::wheelEvent(QWheelEvent * we )
+			int x = we->x();
+
+			if (x >= VALUES_WIDTH) {
+		// set or move value
```
Yes, no problem.

I copied and modified another section of code, and those came along with it.

tecknixia commented on src/gui/editors/AutomationEditor.cpp:
```diff
@@ -1105,7 +1110,11 @@ inline void AutomationEditor::drawCross( QPainter & p )
 	{
-		QToolTip::showText( tt_pos, QString::number( scaledLevel ), this );
+		if (m_pointYLevel == 0) {
+			QToolTip::showText( tt_pos, QString::number( scaledLevel ), this );
```
Ahh, yes, I missed those.

CYBERDEViLNL replied:
You missed line 1727:
```cpp
			float level = getLevel( we->y() );
```

For the rest style looks good now, you may hit resolved after adjusting this :-)

tecknixia commented on src/gui/editors/AutomationEditor.cpp:
```diff
@@ -1709,6 +1718,61 @@ void AutomationEditor::wheelEvent(QWheelEvent * we )
+						// mouse wheel down
+						} else if (we->delta() > 0)	{
+							level = roundf(it.value() * 1000) / 1000 +
+							 m_pattern->firstObject()->step<float>();
```
Sure.

Here again, I was trying to follow conventions and wasn't sure.  Spaces before and after operators.  I suppose, however, the space before the line is more than enough.

CYBERDEViLNL replied:
Yes spaces should be placed before and after operators thats good, but this line is split. 

I'm sorry, made a mistake here. Can you instead of removing the space, replace it with one tab? So split lines are indented. (since we use tabs as indentation) It still fits the max 80 chars per line then.

```cpp
							level = roundf(it.value() * 1000) / 1000 +
								m_pattern->firstObject()->step<float>();
```

tecknixia replied:
No problem.

tecknixia commented:
@CYBERDEViLNL Thank you for the style review.

tecknixia committed:
Style review fixes

tecknixia committed:
Style Review Addition

tecknixia committed:
Fixed Indentation Inconsistencies

CYBERDEViLNL commented:
For updating the tooltip while wheeling you might want to add something like this:

Add to header:
```cpp
	ulong m_pointYLevelTimestamp;
```

Change `mouseMoveEvent`:
```cpp
	if (mouseEvent->timestamp() > m_pointYLevelTimestamp)
	{
		m_pointYLevel = 0;
	}
```

Your `wheelEvent` code:
```cpp
	m_pointYLevelTimestamp = we->timestamp();
```

I don't know other solution for that.

tecknixia committed:
Removed Spaces Within Parameters

tecknixia commented:
Could you please explain?  What is the reason for doing that?  Is it better?

It already updates the tooltip when scrolling, and shows mouse pointer position when mouse is moved, which seems fine to me.

What it does not do is show the tick y level when mouse is over the automation point _before scrolling_, it just shows the pointer y position until the wheel is scrolled.

CYBERDEViLNL commented:
While wheeling over a point the tooltip doesn't get updated here, it's because `wheelEvent` also triggers a `mouseMoveEvent` here and then `m_pointYLevel` is reset to `0`. I'm on Qt version `5.13.1-2` btw.

tecknixia commented:
Ok, I understand now.

I was not familiar with the interaction between wheelEvent and mouseMove.  It works for me (Qt 5.5.1 on Linux Mint 18.3 64-bit), perhaps I should upgrade Qt if there's that much of a difference.

Maybe I can just set `m_pointYLevel` to 0 when the mouse x or y value changes... I'll play around with it and try your method.  Thanks for the suggestion.

CYBERDEViLNL commented:
I also didn't know that a `wheelEvent` did trigger a `mouseMoveEvent`, I haven't found it in the Qt docs either. Maybe it's a Qt bug?

tecknixia commented:
A condition in `mouseMoveEvent` to identify that the mouse pointer is over an automation point would solve this problem, and lay the foundation for resolving the other two issues.  It may be worth the rewrite.  However, it might be better to patch up this PR, because I don't know how long it'll take to rewrite.  I'll do that first.

tecknixia committed:
Tooltip Display Fix

tecknixia commented:
I think I figured out the reason this was never completed in the first place.  I believe that this line:

`else if( mouseEvent->buttons() & Qt::NoButton && m_editMode == DRAW )`

Inside `mouseMoveEvent` was never `true`.  First, I replaced the detection code with the one I was using for the scroll wheel.  Then, since I wasn't getting a result, I moved the contained code to the end of the `if else` as an `else`.

Now I'm getting a cursor change when hovering over an automation point and the test value is showing up on the tooltip.

tecknixia committed:
Rewrite of Pull Request

tecknixia committed:
Merge remote-tracking branch 'upstream/master'

tecknixia commented:
I fixed both the point level on hover and window scroll issues.  Left a `//TODO:` for the future `mouseDoubleClick` enhancement.

About to push a BIG commit...

tecknixia commented on src/gui/editors/AutomationEditor.cpp:
```diff
@@ -1006,14 +943,75 @@ void AutomationEditor::mouseMoveEvent(QMouseEvent * mouseEvent )
+				// sets drawCross tooltip back to mouse y position
+				if (mouseEvent->timestamp() > m_pointYLevelTimestamp)
+				{
+						m_pointYLevel = 0;
```
Of course after all that I _would_ have one extra tab... :-1:

tecknixia committed:
Whitespace fixes

tecknixia commented:
Ok, things should be good at this point.

I would like some feedback.    Let me know if there is anything wrong, code or function.

Is this is functioning properly on other systems, and what do you think about user experience?

tecknixia committed:
Added Double Click Input Dialog Functionality

tecknixia committed:
Engagement Area Enhancement

tecknixia commented:
Two enhancements:

1. Added the ability to enter a value when double clicking on an automation point
2. Adjusted the area to only include functionality near the automation points.

My previous comment still stands.  I would like some feedback.
Let me know if there is anything wrong, code or function.

stevphie commented:
> 
> 
> Two enhancements:
> 
>     1. Added the ability to enter a value when double clicking on an automation point
> 
>     2. Adjusted the area to only include functionality near the automation points.
> 
> 
> My previous comment still stands. I would like some feedback.
> Let me know if there is anything wrong, code or function.

I think the tittle does not reflect to what you just did (pointing at point No. 1)

tecknixia committed:
Mouse Control Enhancements

tecknixia committed:
Merge remote-tracking branch 'upstream/master'

PhysSong commented:
@tecknixia I'm trying to review this pull request, but it's not easy due to some irrelevant reformattings and reorderings. Could you revert those changes to make the review easier?

tecknixia commented:
@PhysSong I started over.  I've closed this one and created two new ones to replace it: #5291  and #5292

PhysSong commented:
@tecknixia Thanks. I will review them soon.
